File: proto-csharp/ProtoMessage/ProtoMessage/ApeMessage/Reflect/MessageFieldDescriptor.cs
using System;
using System.Reflection;
using System.Collections.Generic;
using Google.ProtocolBuffers;
using com.spaceape.protobuf;

namespace com.spaceape.protobuf.reflect
{
  public class MessageFieldDescriptor : FieldDescriptor
  {
    private ConstructorFactory constructorFactory;

    public MessageFieldDescriptor(int fieldNumber, String fieldName, Type clazz, ConstructorFactory constructorFactory, Type fieldClass) : base(fieldNumber, fieldName, clazz)
    {
      this.constructorFactory = constructorFactory;
    }

    override public Object readField(CodedInputStream input, CodeGenContext context)
    {
      uint tag;
      input.ReadTag(out tag);
      return readMessage(input, context, constructorFactory);
    }

    override public void writeField(CodedOutputStream output, CodeGenContext context, Object obj)
    {
      GeneratedMessage msg = (GeneratedMessage)obj;
      writeMessage(output, context, fieldNumber, msg);
    }

    private GeneratedMessage readMessage(CodedInputStream input, CodeGenContext context, ConstructorFactory constructorFactory)
    {
      bool hasValue = input.ReadBool();
      if (!hasValue)
      {
        uint tag;
        input.ReadTag(out tag);
        if (tag != 0 && WireFormat.GetTagWireType(tag) != WireFormat.WireType.EndGroup)
        {
          throw new Exception("invalid null message end with tag " + tag);
        }
        return null;
      }
      String id = input.ReadString();

      if (context.contains(id))
        return context.getMessage(id);
      else
      {
        String className = input.ReadString();

        GeneratedMessage message = constructorFactory.newInstance(className);
        context.touch(id, message);
        foreach (KeyValuePair<int, FieldDescriptor> entry in message.getFieldDescriptors())
        {
          entry.Value.readAndSet(input, context, message);
        }
        return message;
      }
    }

    public void writeMessage(CodedOutputStream output, CodeGenContext context, int fieldNumber, GeneratedMessage message)
    {
      output.WriteTag(fieldNumber, WireFormat.WireType.StartGroup);
      writeMessageNoTag(output, context, message);
      output.WriteTag(fieldNumber, WireFormat.WireType.EndGroup);
    }

    private void writeMessageNoTag(CodedOutputStream output, CodeGenContext context, GeneratedMessage message)
    {
      if (message == null)
      {
        output.WriteBoolNoTag(false);
      }
      else
      {
        output.WriteBoolNoTag(true);
        String id = context.generateMessageId(message);
        output.WriteStringNoTag(id);

        if (!context.contains(id))
        {
          context.touch(id, message);
          output.WriteStringNoTag(message.GetType().FullName);

          foreach (KeyValuePair<int, FieldDescriptor> entry in message.getFieldDescriptors())
          {
            entry.Value.writeField(output, context, field.GetValue(message));
          }
        }
      }
    }
  }
}

File: proto-java/src/main/java/com/spaceape/protobuf/reflect/MessageFieldDescriptor.java
package com.spaceape.protobuf.reflect;

import com.google.protobuf.CodedInputStream;
import com.google.protobuf.CodedOutputStream;
import com.google.protobuf.WireFormat;
import com.spaceape.protobuf.CodeGenContext;
import com.spaceape.protobuf.ConstructorFactory;
import com.spaceape.protobuf.GeneratedMessage;

import java.io.IOException;

public class MessageFieldDescriptor extends FieldDescriptor {
    private ConstructorFactory constructorFactory;

    public MessageFieldDescriptor(int fieldNumber, String fieldName, Class clazz, ConstructorFactory constructorFactory, Class fieldClass){
        super(fieldNumber, fieldName, clazz, fieldClass);
        this.constructorFactory = constructorFactory;
    }

    public GeneratedMessage readField(CodedInputStream input, CodeGenContext context) throws IOException {
        return readMessage(input, context, constructorFactory);
    }

    public void writeField(CodedOutputStream output, CodeGenContext context, Object object) throws IOException{
        GeneratedMessage msg = (GeneratedMessage)object;
        writeMessage(output, context, fieldNumber, msg);
    }

    private GeneratedMessage readMessage(CodedInputStream input, CodeGenContext context, ConstructorFactory constructorFactory) throws IOException{
        final boolean hasValue = input.readBool();
        if (!hasValue){
            int tag = input.readTag();
            if (tag != 0 && WireFormat.getTagWireType(tag) != WireFormat.WIRETYPE_END_GROUP){
                throw new IllegalArgumentException("invalid null message end with tag "+tag);
            }
            return null;
        }
        String id = input.readStringRequireUtf8();

        if (context.contains(id)) return context.getMessage(id);
        else {
            String className = input.readStringRequireUtf8();

            GeneratedMessage message = constructorFactory.newInstance(className);
            context.touch(id, message);
            return GenerateMessageStream.read(input, context, message);
        }
    }

    public void writeMessage (CodedOutputStream output, CodeGenContext context, int fieldNumber, GeneratedMessage message) throws IOException{
        output.writeTag(fieldNumber, WireFormat.WIRETYPE_START_GROUP);
        writeMessageNoTag(output, context, message);
        output.writeTag(fieldNumber, WireFormat.WIRETYPE_END_GROUP);
    }

    public static void writeMessageNoTag (CodedOutputStream output, CodeGenContext context, GeneratedMessage message) throws IOException{
        if (message == null) {
            output.writeBoolNoTag(false);
        } else {
            output.writeBoolNoTag(true);
            String id = context.generateMessageId(message);
            output.writeStringNoTag(id);

            if (!context.contains(id)) {
                context.touch(id, message);
                output.writeStringNoTag(message.getClass().getName());

                try {
                    for (FieldDescriptor descriptor : message.getFieldDescriptors().values()) {
                        descriptor.writeField(output, context, descriptor.field.get(message));
                    }
                }catch (Exception e){
                    throw new RuntimeException(e);
                }
            }
        }
    }
}
